### PR TITLE
Add API v2 events admin

### PR DESCRIPTION
**Pull request by se-bastiaan**
Part of #1428

### Summary
This introduces full functionality to manage events via the v2 API.

### How to test
Steps to test the changes you made:
1. Go to the API docs on /api/docs and select v2.
2. Check the endpoints and try some things out

### CODE REVIEW

**Comment by JobDoesburg** [comment]
deploy review

**Comment by DeD1rk** [comment]
@JobDoesburg GitHub actions is a little bit broken; it sometimes skips the deploy/remove review actions

**Comment by JobDoesburg** [comment]
> @JobDoesburg GitHub actions is a little bit broken; it sometimes skips the deploy/remove review actions

Yeah I notice that. Don't see how though

File: website/events/api/v2/admin/views.py
Comment (JobDoesburg) [inline website/events/api/v2/admin/views.py:28]: This API view triggers 205 SQL queries on my db. A prefetch on the activemembergroup would be nice

**Comment by JobDoesburg** [comment]
Shouldn't these API views also have a Django model permission permission class?

**Comment by se-bastiaan** [comment]
> 
> 
> Shouldn't these API views also have a Django model permission permission class?

I tihnk you asked this before already, have you seen the AdminXAPIView base classes?

**Comment by JobDoesburg** [comment]
> > Shouldn't these API views also have a Django model permission permission class?
> 
> I tihnk you asked this before already, have you seen the AdminXAPIView base classes?

Alright, yeah I also thought I already asked that but I couldn't find the message in this thread anymore haha

File: website/events/api/v2/filters.py
Comment (se-bastiaan) [reply website/events/api/v2/filters.py:46]: No you should just not send spaces, that would be stupid

File: website/events/models/event.py
Comment (se-bastiaan) [reply website/events/models/event.py:445]: Which variables? The ones for the reminder still have the language field, those are not from this app.

File: website/events/api/v2/admin/views.py
Comment (se-bastiaan) [reply website/events/api/v2/admin/views.py:28]: goddamn good idea

image
image

File: website/events/api/v2/admin/serializers/event_registration.py
Comment (JobDoesburg) [inline website/events/api/v2/admin/serializers/event_registration.py:31]: Going to `localhost:8000/api/v2/admin/events/11/registrations/7/` shows me the ability to `PUT` to a member profile, even if I do not have permissions for this. Maybe the MemberSerializer should have `readonly=True`?

File: website/events/api/v2/admin/views.py
Comment (JobDoesburg) [inline website/events/api/v2/admin/views.py:29]: Still quite a hell lot of queries that are not really necessary

image

Can probably be fixed with prefetching the event registrations as well. But we could also fix it by making the num_participants a queryable property which I'm more in favor of (#1617 which should probably be renamed 'queryable properties for active members and events'). So you can also ignore it here if you want

File: website/events/api/v2/admin/views.py
Comment (JobDoesburg) [inline website/events/api/v2/admin/views.py:50]: I think the DetailAPIView shouldn't be a CreateAPIView, only UpdateAPIView. The ListAPIView should be a CreateAPIView in my opinion,

File: website/events/api/v2/admin/serializers/event.py
Comment (JobDoesburg) [inline website/events/api/v2/admin/serializers/event.py:74]: No DecimalFields, like we use in payments and sales? I think DecimalFields are better, as that's what we're using in our database as well (with `decimal_places=2, max_digits=6`)

**Review by JobDoesburg — CHANGES_REQUESTED** [review]

**Comment by JobDoesburg** [comment]
After this has been merged, I should redo the `sales` API to also use the `AdminAPIViews` and everything related

File: website/events/api/v2/admin/serializers/event_registration.py
Comment (JobDoesburg) [reply website/events/api/v2/admin/serializers/event_registration.py:31]: Idem for the organizer, btw

File: website/events/api/v2/admin/serializers/event_registration.py
Comment (JobDoesburg) [inline website/events/api/v2/admin/serializers/event_registration.py:19]: The payment should be read only as payments will be managed by the payments API

File: website/events/api/v2/admin/views.py
Comment (se-bastiaan) [reply website/events/api/v2/admin/views.py:29]: I tried adding the `eventregistration_set` to the prefetch but that added a query instead of removing one. Which is understandable since the queries you see here are the registration information field which are completely independent models. Not sure I can prefetch those through the event registrations.

Do agree that queryable properties are probably nice, but really out of scope for this PR.

File: website/events/api/v2/admin/serializers/event_registration.py
Comment (se-bastiaan) [reply website/events/api/v2/admin/serializers/event_registration.py:31]: That's only because those views assume that the output of a serializer is exactly the same as the input of a serializer. Which I changed in this case, since it accepts a different member id and not a full object. Don't think I can do anything to prevent that. And marking the field as read only could probably make it impossible to edit the field at all.

File: website/events/api/v2/admin/views.py
Comment (se-bastiaan) [reply website/events/api/v2/admin/views.py:50]: That would overcomplicate the other view, which shows a list. This one is for single objects and editing those. And it uses a different serializer. Thus I chose to include it here instead.

File: website/events/api/v2/admin/serializers/event_registration.py
Comment (se-bastiaan) [inline website/events/api/v2/admin/serializers/event_registration.py:22]:
```suggestion
        fields = (
            "pk",
            "present",
            "queue_position",
            "date",
            "date_cancelled",
            "payment",
            "member",
            "name",
        )
        read_only_fields = ("payment",)
```

File: website/events/api/v2/admin/views.py
Comment (JobDoesburg) [reply website/events/api/v2/admin/views.py:50]: But now, if you want to create a new event, you must do that via another event, also using the `id` of a different event.

I don't think it really makes the other view too complicated

File: website/events/api/v2/admin/serializers/event_registration.py
Comment (se-bastiaan) [reply website/events/api/v2/admin/serializers/event_registration.py]: Looks like it still works when I make them read only :)

File: website/events/api/v2/admin/views.py
Comment (se-bastiaan) [reply website/events/api/v2/admin/views.py]: I was dumb

File: website/events/api/v2/admin/serializers/event.py
Comment (JobDoesburg) [inline website/events/api/v2/admin/serializers/event.py:25]: This should also be DecimaFields

File: website/events/api/v2/admin/views.py
Comment (JobDoesburg) [inline website/events/api/v2/admin/views.py:96]: This should be an AdminAPIView I think?

also maybe the `get` `put` and `patch` methods should call their `super()` methods

**Review by JobDoesburg — CHANGES_REQUESTED** [review]

**Comment by JobDoesburg** [comment]
Looks good, will approve after these 2 changes

File: website/events/api/v2/admin/views.py
Comment (se-bastiaan) [reply website/events/api/v2/admin/views.py]: There are no super methods. But I had to include the permission mixin indeed.

**Review by JobDoesburg — APPROVED** [review]

**Pull request merged**
